src/v1/parser.rs

```diff
@@ -123,15 +123,105 @@ quick_error! {
+/// Thus `File("/b") < Dir("/a")`.
+#[derive(Debug, PartialEq, Eq)]
 pub enum EntryKind<P: AsRef<Path>> {
     /// A directory
     Dir(P),
     /// A file or a symbolic link
     File(P),
 }
 
+impl<P: AsRef<Path>> EntryKind<P> {
+    /// Get path of the entry
+    pub fn path(&self) -> &Path {
+        use self::EntryKind::*;
+        match *self {
+            Dir(ref p) | File(ref p) => p.as_ref(),
+        }
+    }
+
+    /// Converts to `EntryKind<&Path>`
+    pub fn as_ref(&self) -> EntryKind<&Path> {
+        use self::EntryKind::*;
+        match *self {
+            Dir(ref p) => Dir(p.as_ref()),
+            File(ref p) => File(p.as_ref()),
+        }
+    }
+
+    /// Clones path and returns `EntryKind<PathBuf>`
+    pub fn cloned(&self) -> EntryKind<PathBuf> {
+        use self::EntryKind::*;
+        match *self {
+            Dir(ref p) => Dir(p.as_ref().to_path_buf()),
+            File(ref p) => File(p.as_ref().to_path_buf()),
+        }
+    }
+}
+
+impl<P> PartialOrd for EntryKind<P>
+    where P: AsRef<Path> + PartialEq + Eq
+{
+    fn partial_cmp(&self, other: &EntryKind<P>)
+        -> Option<Ordering>
+    {
+        Some(self.cmp(other))
+    }
+}
+
+impl<P> Ord for EntryKind<P>
+    where P: AsRef<Path> + PartialEq + Eq
+{
+    fn cmp(&self, other: &EntryKind<P>) -> Ordering {
+        use std::cmp::Ordering::*;
+        use self::EntryKind::*;
+
+        // we cannot just compare paths since directories and files are placed
+        // differently in a signature file
+        match *self {
+            Dir(ref path) => {
+                match *other {
+                    Dir(ref other_path) => {
+                        path.as_ref().cmp(other_path.as_ref())
+                    },
+                    File(ref other_path) => {
+                        let other_parent = other_path.as_ref().parent()
+                            .unwrap_or(Path::new("/"));
+                        match path.as_ref().cmp(other_parent.as_ref()) {
+                            Less | Equal => Less,
+                            Greater => Greater,
+                        }
+                    }
+                }
+            },
+            File(ref path) => {
+                let parent = path.as_ref().parent().unwrap_or(Path::new("/"));
+                match *other {
+                    Dir(ref other_path) => {
+                        match parent.cmp(other_path.as_ref()) {
+                            Less => Less,
+                            Greater | Equal => Greater,
+                        }
+                    },
+                    File(ref other_path) => {
+                        let other_parent = other_path.as_ref().parent()
+                            .unwrap_or(Path::new("/"));
+                        match parent.cmp(other_parent) {
+                            Less => Less,
+                            Greater => Greater,
+                            Equal => path.as_ref().cmp(other_path.as_ref()),
```

Comment:
Well, it's impossible to represent empty directory (only root), because that would be an empty line. And it's impossible to represent File that has an empty filename, because we would parse size as a filename in this case.

So I believe we can assert on it.